### PR TITLE
fix(serialization): recording serialization properly converts recording state types

### LINKED ISSUE #313: [Bug] JMC and `jdk.jfr` models of `RecordingState` differ
https://github.com/cryostatio/cryostat-core/blob/e77984d5af9cd614ef3cd582264636353d50551c/src/main/java/io/cryostat/core/serialization/SerializableRecordingDescriptor.java#L70

There is a latent bug here: these two `RecordingState` types from the JMC type hierarchy and from `jdk.jfr.Recording` use two different enums with differently named enum member values to represent states a Flight Recording may be in. Using a simple name-based `valueOf()` call is unsafe since the names do not match `1:1`, nor is one set a subset of the other.

### PR DESCRIPTION
**Pull request by andrewazores**
Fixes #313

### CODE REVIEW

File: src/main/java/io/cryostat/core/serialization/SerializableRecordingDescriptor.java [hunk 1]
```diff
@@ -101,6 +101,25 @@ public SerializableRecordingDescriptor(SerializableRecordingDescriptor o) {
...
+            case NEW:
+                return RecordingState.CREATED;
+            case DELAYED:
+                return RecordingState.CREATED;
+            case RUNNING:
+                return RecordingState.RUNNING;
+            case STOPPED:
+                return RecordingState.STOPPED;
+            case CLOSED:
+                return RecordingState.STOPPED;
```
Comment (mwangggg) [inline src/main/java/io/cryostat/core/serialization/SerializableRecordingDescriptor.java:115]: should this be RecordingState.CLOSED?

**Comment by andrewazores** [comment]
```
 [INFO] Scanning for projects...
[INFO] 
[INFO] ------------------------< io.cryostat:cryostat >------------------------
[INFO] Building cryostat 2.5.0-SNAPSHOT
[INFO]   from pom.xml
[INFO] --------------------------------[ jar ]---------------------------------
[INFO] Downloading from github: https://maven.pkg.github.com/cryostatio/cryostat-core/io/cryostat/cryostat-core/2.26.0-SNAPSHOT/maven-metadata.xml
[INFO] Downloading from github: https://maven.pkg.github.com/cryostatio/cryostat-core/io/cryostat/cryostat-core/2.26.0-SNAPSHOT/cryostat-core-2.26.0-SNAPSHOT.pom
Warning:  The POM for io.cryostat:cryostat-core:jar:2.26.0-SNAPSHOT is missing, no dependency information available
[INFO] Downloading from github: https://maven.pkg.github.com/cryostatio/cryostat-core/io/cryostat/cryostat-core/2.26.0-SNAPSHOT/cryostat-core-2.26.0-SNAPSHOT.jar
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  1.642 s
[INFO] Finished at: 2023-12-04T18:45:26Z
[INFO] ------------------------------------------------------------------------
Error:  Failed to execute goal on project cryostat: Could not resolve dependencies for project io.cryostat:cryostat:jar:2.5.0-SNAPSHOT: Could not find artifact io.cryostat:cryostat-core:jar:2.26.0-SNAPSHOT in github (https://maven.pkg.github.com/cryostatio/cryostat-core) -> [Help 1]
Error:  
Error:  To see the full stack trace of the errors, re-run Maven with the -e switch.
Error:  Re-run Maven using the -X switch to enable full debug logging.
Error:  
Error:  For more information about the errors and possible solutions, please read the following articles:
Error:  [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/DependencyResolutionException
Error: Process completed with exit code 1.
```

CI failure looks like the builder is trying to use `2.26.0-SNAPSHOT` for `-core` because that's the version in the `pom.xml` in this PR (since that's what is on `main`), but that is not a published version since we only publish actual releases, not snapshots.

I think the CI builder needs to build `-core` to test it, then upload that artifact. The "build Cryostat" task should then download that artifact again and re-install it into the local `.m2` so that the Cryostat image build can find it and use it, so that the Cryostat build and tests use a `-core` artifact that is derived from the PR being tested.

**Comment by mwangggg** [comment]
https://github.com/cryostatio/cryostat-core/blob/e77984d5af9cd614ef3cd582264636353d50551c/src/main/java/org/openjdk/jmc/rjmx/services/jfr/internal/RecordingDescriptorV2.java#L91-L104 I was taking a look at some other places where the recording states are mentioned and I was just wondering where these states come from? I know they are composite data from the recordings, but they aren't entirely the `jdk.jfr.recording` or the `RecordingStates` sets?

**Comment by mwangggg** [comment]
> I think the CI builder needs to build `-core` to test it, then upload that artifact. The "build Cryostat" task should then download that artifact again and re-install it into the local `.m2` so that the Cryostat image build can find it and use it, so that the Cryostat build and tests use a `-core` artifact that is derived from the PR being tested.

This is what I intended when I first implemented the Cryostat build testing, but I think there's an error caching or retrieving the cached files. I'll take a look at it right now.

**Comment by andrewazores** [comment]
Basically I painted us into a corner a long time ago. We probably want to move in the direction of just using the `jdk.jfr` classes to model our data as much as possible, but a lot of our stack is built on top of JMC and its `org.openjdk.jmc.rjmx.services.jfr` classes, which are roughly equivalent but do have some differences.

One nice thing about the JMC stuff is that it provides nicer type signatures in many cases that make it clearer/less error-prone what you're doing - though sometimes this is a little boilerplate-y.

On the downside, those classes and the various Unit types don't always serialize well, and there is at least this one case where the state of a recording is modelled differently between the two packages. I don't know if that recording state difference is because the JMC side of it is more geared toward JMC's UI display, or if perhaps it's an older state model that looks more like the old V1 proprietary JFR implementation rather than the V2 OpenJDK JFR implementation. Regardless of the reason for the difference, we would probably be better off migrating away from using JMC's abstractions except in cases where we really want/need it, like for Automated Analysis. But that will be a pretty big undertaking to refactor all of that and test for any regressions (in serialization especially), and it's just as cleanup/tech-debt chore rather than feature work, so it hasn't been very high priority.

**Review by mwangggg — APPROVED** [review]
LGTM

**Comment by aptmac** [comment]
> Looks to me like the JMC RecordingState model is indeed modelled after the old V1 JFR implementation, and they just mapped the OpenJDK JFR V2 states onto the existing model, which we've inherited from JMC.

`IRecordingDescriptor` and `RecordingDescriptorV2` haven't been updated since JMC was open-sourced. Perhaps it's time to update the `RecordingState` enum and touch-up `RecordingDescriptorV2` at the same time.

**Comment by andrewazores** [comment]
https://github.com/cryostatio/cryostat-core/actions/runs/7105513277

**Comment by aali309** [comment]
Did you see this @andrewazores 
```Error: cannot create directory ‘/home/runner/work/cryostat-core/cryostat-core/target/[ERROR] Failed to execute goal org.codehaus.mojo:build-helper-maven-plugin:3.5.0:regex-property (image-tag-to-lower) on project cryostat: Execution image-tag-to-lower of goal org.codehaus.mojo:build-helper-maven-plugin:3.5.0:regex-property failed: Plugin org.codehaus.mojo:build-helper-maven-plugin:3.5.0 or one of its dependencies could not be resolved: Cannot access central (https:’: File name too long
Error: cannot create directory ‘/home/runner/work/cryostat-core/cryostat-core/target/[ERROR] Failed to execute goal org.codehaus.mojo:build-helper-maven-plugin:3.5.0:regex-property (image-tag-to-lower) on project cryostat: Execution image-tag-to-lower of goal org.codehaus.mojo:build-helper-maven-plugin:3.5.0:regex-property failed: Plugin org.codehaus.mojo:build-helper-maven-plugin:3.5.0 or one of its dependencies could not be resolved: Cannot access central (https:’: File name too long
Error: d-integration-tests.bash: line 95: /home/runner/work/cryostat-core/cryostat-core/target/[ERROR] Failed to execute goal org.codehaus.mojo:build-helper-maven-plugin:3.5.0:regex-property (image-tag-to-lower) on project cryostat: Execution image-tag-to-lower of goal org.codehaus.mojo:build-helper-maven-plugin:3.5.0:regex-property failed: Plugin org.codehaus.mojo:build-helper-maven-plugin:3.5.0 or one of its dependencies could not be resolved: Cannot access central (https://repo.maven.apache.org/maven2) in offline mode and the artifact org.apache-extras.beanshell:bsh:jar:2.0b6 has not been downloaded from it before. -> [Help 1]
Error:  
Error:  To see the full stack trace of the errors, re-run Maven with the -e switch.
Error:  Re-run Maven using the -X switch to enable full debug logging.
Error:  
Error:  For more information about the errors and possible solutions, please read the following articles:
Error:  [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/PluginResolutionException-2023-12-05T19:19+00:00.client.log: File name too long
Error: d-integration-tests.bash: line 96: /home/runner/work/cryostat-core/cryostat-core/target/[ERROR] Failed to execute goal org.codehaus.mojo:build-helper-maven-plugin:3.5.0:regex-property (image-tag-to-lower) on project cryostat: Execution image-tag-to-lower of goal org.codehaus.mojo:build-helper-maven-plugin:3.5.0:regex-property failed: Plugin org.codehaus.mojo:build-helper-maven-plugin:3.5.0 or one of its dependencies could not be resolved: Cannot access central (https://repo.maven.apache.org/maven2) in offline mode and the artifact org.apache-extras.beanshell:bsh:jar:2.0b6 has not been downloaded from it before. -> [Help 1]
Error:  
Error:  To see the full stack trace of the errors, re-run Maven with the -e switch.
Error:  Re-run Maven using the -X switch to enable full debug logging.
Error:  
Error:  For more information about the errors and possible solutions, please read the following articles:
Error:  [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/PluginResolutionException-2023-12-05T19:19+00:00.server.log: File name too long
Error: d-integration-tests.bash: line 97: /home/runner/work/cryostat-core/cryostat-core/target/[ERROR] Failed to execute goal org.codehaus.mojo:build-helper-maven-plugin:3.5.0:regex-property (image-tag-to-lower) on project cryostat: Execution image-tag-to-lower of goal org.codehaus.mojo:build-helper-maven-plugin:3.5.0:regex-property failed: Plugin org.codehaus.mojo:build-helper-maven-plugin:3.5.0 or one of its dependencies could not be resolved: Cannot access central (https://repo.maven.apache.org/maven2) in offline mode and the artifact org.apache-extras.beanshell:bsh:jar:2.0b6 has not been downloaded from it before. -> [Help 1]
Error:  
Error:  To see the full stack trace of the errors, re-run Maven with the -e switch.
Error:  Re-run Maven using the -X switch to enable full debug logging.
Error:  
Error:  For more information about the errors and possible solutions, please read the following articles:
Error:  [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/PluginResolutionException-2023-12-05T19:19+00:00.client.log: File name too long```

**Comment by andrewazores** [comment]
No, I didn't notice that. Where is it from?

**Comment by aali309** [comment]
> No, I didn't notice that. Where is it from?



> Cryostat Test: All tests pass ✅ https://github.com/cryostatio/cryostat-core/actions/runs/7105513277

Here..

**Comment by andrewazores** [comment]
Ah, I see. Odd that the workflow step succeeded despite those errors. Despite the errors it looks like it still did run the tests successfully, too...

**Comment by aali309** [comment]
Yes, was wondering  the same how did we get to receive the `All tests pass`  comment.

**Pull request merged**